### ISSUE TITLE
installing levior in virtualenv + service mode command line argument is ignored

### ISSUE THREAD

**Issue by norayr** [pinnaculum/levior#1]
i have created venv:
```
[2023-11-19 04:28:54] $ virtualenv ~/pip/levior
created virtual environment CPython3.11.6.final.0-64 in 231ms
  creator CPython3Posix(dest=/home/user/pip/levior, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/user/.local/share/virtualenv)
    added seed packages: pip==23.3.1, setuptools==68.2.2, wheel==0.41.2
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
```

activated it:

```
[2023-11-19 04:28:59] $ source ~/pip/levior/bin/activate
```
i tried to install levior:

```
(levior) [2023-11-19 04:29:25] $ cd ~/pip/levior/
(levior) [2023-11-19 04:29:46] $ git clone https://github.com/pinnaculum/levior
Cloning into 'levior'...
remote: Enumerating objects: 114, done.
remote: Counting objects: 100% (114/114), done.
remote: Compressing objects: 100% (54/54), done.
remote: Total 114 (delta 60), reused 105 (delta 51), pack-reused 0
Receiving objects: 100% (114/114), 69.08 KiB | 1.11 MiB/s, done.
Resolving deltas: 100% (60/60), done.
(levior) [2023-11-19 04:29:54] $ pwd
/u/user/pip/levior
(levior) [2023-11-19 04:29:56] $ ls
bin  levior  lib  pyvenv.cfg
(levior) [2023-11-19 04:30:11] $ pip install -e .
Obtaining file:///home/user/pip/levior/levior
  Preparing metadata (setup.py) ... done
Collecting aiogemini@ git+https://github.com/keis/aiogemini.git@master (from levior==1.1.0)
  Cloning https://github.com/keis/aiogemini.git (to revision master) to /tmp/pip-install-07v_jxhj/aiogemini_99061c20f1fa40bdaa390ee6474513f5
  Running command git clone --filter=blob:none --quiet https://github.com/keis/aiogemini.git /tmp/pip-install-07v_jxhj/aiogemini_99061c20f1fa40bdaa390ee6474513f5
  Resolved https://github.com/keis/aiogemini.git to commit c3a36bf6933f98d832aac0daaa840146037a375c
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Collecting aiohttp>=3.8.1 (from levior==1.1.0)
  Using cached aiohttp-3.9.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (7.4 kB)
Collecting aiohttp-socks>=0.6.0 (from levior==1.1.0)
  Using cached aiohttp_socks-0.8.4-py3-none-any.whl.metadata (3.1 kB)
Collecting diskcache>=5.4.0 (from levior==1.1.0)
  Using cached diskcache-5.6.3-py3-none-any.whl.metadata (20 kB)
Collecting easydict>=1.9 (from levior==1.1.0)
  Using cached easydict-1.11-py3-none-any.whl
Collecting yarl>=1.7.2 (from levior==1.1.0)
  Using cached yarl-1.9.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (282 kB)
Collecting markdownify>=0.11.2 (from levior==1.1.0)
  Using cached markdownify-0.11.6-py3-none-any.whl (16 kB)
Collecting md2gemini>=1.9.0 (from levior==1.1.0)
  Using cached md2gemini-1.9.1-py3-none-any.whl.metadata (10 kB)
Collecting multidict<5.0,>=4.5 (from levior==1.1.0)
  Using cached multidict-4.7.6-cp311-cp311-linux_x86_64.whl
Collecting PyYAML>=6.0 (from levior==1.1.0)
  Using cached PyYAML-6.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.1 kB)
Collecting attrs>=17.3.0 (from aiohttp>=3.8.1->levior==1.1.0)
  Using cached attrs-23.1.0-py3-none-any.whl (61 kB)
Collecting frozenlist>=1.1.1 (from aiohttp>=3.8.1->levior==1.1.0)
  Using cached frozenlist-1.4.0-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.2 kB)
Collecting aiosignal>=1.1.2 (from aiohttp>=3.8.1->levior==1.1.0)
  Using cached aiosignal-1.3.1-py3-none-any.whl (7.6 kB)
Collecting python-socks<3.0.0,>=2.4.3 (from python-socks[asyncio]<3.0.0,>=2.4.3->aiohttp-socks>=0.6.0->levior==1.1.0)
  Using cached python_socks-2.4.3-py3-none-any.whl.metadata (7.1 kB)
Collecting beautifulsoup4<5,>=4.9 (from markdownify>=0.11.2->levior==1.1.0)
  Using cached beautifulsoup4-4.12.2-py3-none-any.whl (142 kB)
Collecting six<2,>=1.15 (from markdownify>=0.11.2->levior==1.1.0)
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting mistune<3,>=2.0.0 (from md2gemini>=1.9.0->levior==1.1.0)
  Using cached mistune-2.0.5-py2.py3-none-any.whl (24 kB)
Collecting cjkwrap (from md2gemini>=1.9.0->levior==1.1.0)
  Using cached CJKwrap-2.2-py2.py3-none-any.whl (4.3 kB)
Collecting wcwidth (from md2gemini>=1.9.0->levior==1.1.0)
  Using cached wcwidth-0.2.10-py2.py3-none-any.whl.metadata (14 kB)
Collecting idna>=2.0 (from yarl>=1.7.2->levior==1.1.0)
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting cryptography<42.0,>=3.2 (from aiogemini@ git+https://github.com/keis/aiogemini.git@master->levior==1.1.0)
  Using cached cryptography-41.0.5-cp37-abi3-manylinux_2_28_x86_64.whl.metadata (5.2 kB)
Collecting soupsieve>1.2 (from beautifulsoup4<5,>=4.9->markdownify>=0.11.2->levior==1.1.0)
  Using cached soupsieve-2.5-py3-none-any.whl.metadata (4.7 kB)
Collecting cffi>=1.12 (from cryptography<42.0,>=3.2->aiogemini@ git+https://github.com/keis/aiogemini.git@master->levior==1.1.0)
  Using cached cffi-1.16.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.5 kB)
Collecting async-timeout>=3.0.1 (from python-socks[asyncio]<3.0.0,>=2.4.3->aiohttp-socks>=0.6.0->levior==1.1.0)
  Using cached async_timeout-4.0.3-py3-none-any.whl.metadata (4.2 kB)
Collecting pycparser (from cffi>=1.12->cryptography<42.0,>=3.2->aiogemini@ git+https://github.com/keis/aiogemini.git@master->levior==1.1.0)
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Using cached aiohttp-3.9.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)
Using cached aiohttp_socks-0.8.4-py3-none-any.whl (9.6 kB)
Using cached diskcache-5.6.3-py3-none-any.whl (45 kB)
Using cached md2gemini-1.9.1-py3-none-any.whl (26 kB)
Using cached PyYAML-6.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (757 kB)
Using cached cryptography-41.0.5-cp37-abi3-manylinux_2_28_x86_64.whl (4.4 MB)
Using cached frozenlist-1.4.0-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (250 kB)
Using cached python_socks-2.4.3-py3-none-any.whl (52 kB)
Using cached wcwidth-0.2.10-py2.py3-none-any.whl (105 kB)
Using cached async_timeout-4.0.3-py3-none-any.whl (5.7 kB)
Using cached cffi-1.16.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (464 kB)
Using cached soupsieve-2.5-py3-none-any.whl (36 kB)
Building wheels for collected packages: aiogemini
  Building wheel for aiogemini (pyproject.toml) ... done
  Created wheel for aiogemini: filename=aiogemini-0.2.0-py3-none-any.whl size=9801 sha256=abf43f525ccea7dcf50ed947effe2417af73c4d65ec90a39b2a1665d22b25622
  Stored in directory: /tmp/pip-ephem-wheel-cache-ar9nyoh5/wheels/d4/3f/23/08fe4ba10cc9226507248c90a77c376878cbf457348a1f20bb
Successfully built aiogemini
Installing collected packages: wcwidth, python-socks, mistune, easydict, cjkwrap, soupsieve, six, PyYAML, pycparser, multidict, md2gemini, idna, frozenlist, diskcache, attrs, async-timeout, yarl, cffi, beautifulsoup4, aiosignal, markdownify, cryptography, aiohttp, aiohttp-socks, aiogemini, levior
  Running setup.py develop for levior
Successfully installed PyYAML-6.0.1 aiogemini-0.2.0 aiohttp-3.9.0 aiohttp-socks-0.8.4 aiosignal-1.3.1 async-timeout-4.0.3 attrs-23.1.0 beautifulsoup4-4.12.2 cffi-1.16.0 cjkwrap-2.2 cryptography-41.0.5 diskcache-5.6.3 easydict-1.11 frozenlist-1.4.0 idna-3.4 levior-1.1.0 markdownify-0.11.6 md2gemini-1.9.1 mistune-2.0.5 multidict-4.7.6 pycparser-2.21 python-socks-2.4.3 six-1.16.0 soupsieve-2.5 wcwidth-0.2.10 yarl-1.9.2
(levior) [2023-11-19 04:30:36] $ 
```

i can see that packages are installed in `~/pip/levior/lib/python3.11/site-packages`.

now trying to run it:

```
(levior) [2023-11-19 04:30:39] $ cd bin
bash: cd: bin: No such file or directory
(levior) [2023-11-19 04:30:43] $ cd ../bin/
(levior) [2023-11-19 04:30:46] $ ls
activate       activate.ps1      md2gemini  pip3.11     wheel
activate.csh   activate_this.py  pip        python      wheel3
activate.fish  levior            pip3       python3     wheel-3.11
activate.nu    markdownify       pip-3.11   python3.11  wheel3.11
(levior) [2023-11-19 04:30:46] $ cp ../
bin/        .gitignore  levior/     lib/        pyvenv.cfg  
(levior) [2023-11-19 04:30:46] $ cp ../levior/examples/levior.yaml .
(levior) [2023-11-19 04:31:02] $ ./levior -c levior.yaml 
Traceback (most recent call last):
  File "/home/user/pip/levior/bin/./levior", line 33, in <module>
    sys.exit(load_entry_point('levior', 'console_scripts', 'levior')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/pip/levior/bin/./levior", line 25, in importlib_load_entry_point
    return next(matches).load()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/metadata/__init__.py", line 202, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/user/pip/levior/levior/levior/entrypoint.py", line 5, in <module>
    from aiogemini.security import TOFUContext
ModuleNotFoundError: No module named 'aiogemini.security'
(levior) [2023-11-19 04:31:09] $
```

i didn't find word 'security' in any file in `site-packages/aiogemini` by grepping there.

what would you suggest?

**Comment by pinnaculum** [comment]
Hello @norayr.

Just saw that the changes recently pushed in the [gitlab levior repo](https://gitlab.com/cipres/levior) had not been pushed to the GH repo, sorry about that. I just synced it so this will solve the issue of the old aiogemini version code.

I just wanted to comment on the way you setup the virtualenv. When you "source" a virtualenv, it will add the virtualenv's "bin" directory to $PATH. So once it's installed you don't need to chdir to the venv's "bin" directory to run the program, running "levior" (whatever the current directory you're in) will find the program in the "bin" directory because it's in the $PATH. I suggest setting it up like this:

```
git clone https://github.com/pinnaculum/levior
cd levior
python -m venv venv
source venv/bin/activate
pip install .
levior
```

There's also an [AppImage](https://gitlab.com/cipres/levior/-/releases/continuous-master/downloads/levior-latest-x86_64.AppImage) available just in case.

**Comment by norayr** [comment]
thank you!

i was using the appimage, i wanted to build the source myself now.
and thank you for telling about the path.

now, i am running levior with
```
./levior -c levior.yaml --mode=http-proxy
```
or

```
./levior --mode=http-proxy
```

in both cases, if i enter

```
https://lite.duckduckgo.com/lite/?q=test
```

i get a long trace, which ends with:

```
File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
 raise ClientConnectorError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host lite:80 ssl:default [Name or service not known]
```
so it looks like it corrupts the string and somehow duckduckgo.com gets stripped.

if i use http instead of https it's the same:

```
File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 907, in _create_connection
 _, proto = await self._create_direct_connection(req, traces, timeout)
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
 raise ClientConnectorError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host lite:80 ssl:default [Name or service not known]
```

if i go to

```
gemini://localhost/https:/sdf.org
```

then anyway i get

```
File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
 raise ClientConnectorError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host https:80 ssl:default [Name or service not known]
```

however if i call the latest appimage with --mode=http-proxy everything works very well.

**Comment by pinnaculum** [comment]
> ./levior --mode=http-proxy
> ```
> 
> in both cases, if i enter
> 
> ```
> https://lite.duckduckgo.com/lite/?q=test
> ```
> 
> i get a long trace, which ends with:
> 
> ```
> File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
>  raise ClientConnectorError(req.connection_key, exc) from exc
> aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host lite:80 ssl:default [Name or service not known]
> ```
> 
> so it looks like it corrupts the string and somehow duckduckgo.com gets stripped.

Indeed, that's what seems to happen, but i can't reproduce it. Which gemini browser are you using, probably Lagrange ? In the "Network" settings, the HTTP proxy is set to "127.0.0.1:1965" or "localhost:1965" ?

I had already tried lite.duckduckgo.com with lagrange some time ago and it was working back then, and i just tried again and it works fine:

> levior: Serving in http proxy mode
> Processing request Request(url=URL('https://lite.duckduckgo.com/'))
> Processing request Request(url=URL('https://lite.duckduckgo.com/lite/?q=test'))
> Processing request Request(url=URL('https://test-english.com/'))

I just looked at the [fetch code in proxy mode](https://github.com/pinnaculum/levior/blob/master/levior/handler.py#L213) and it's not clear how it could end up stripping anything from the original URL, because levior passes *req.url* to the crawler, which is the exact URL passed by the gemini browser on the socket ...

It would be interesting to edit the code and add a *print(req.url)* statement before the *crawler.fetch()* call to see what's happening. You don't even see any "Processing request ..." logs in the terminal ?

I will try levior with python 3.11 to have the same environment as you have, would be surprised if that's what causes the problem.

> 
> if i use http instead of https it's the same:
> 
> ```
> File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 907, in _create_connection
>  _, proto = await self._create_direct_connection(req, traces, timeout)
> 	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>  File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
>  raise ClientConnectorError(req.connection_key, exc) from exc
> aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host lite:80 ssl:default [Name or service not known]
> ```
> 
> if i go to
> 
> ```
> gemini://localhost/https:/sdf.org
> ```
> 
> then anyway i get
> 
> ```
> File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
>  raise ClientConnectorError(req.connection_key, exc) from exc
> aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host https:80 ssl:default [Name or service not known]
> ```

If you're running levior in *proxy* mode, you **cannot** use *gemini* URLs, it should actually reply in the browser: "Unsupported URL scheme: gemini".

If you're running levior in *server* mode, the correct URL would be:

> gemini://localhost/sdf.org

(you don't pass the full URL but just the domain and path)

The service modes and the URL formats are documented [here](https://github.com/pinnaculum/levior/tree/master#service-modes)

> 
> however if i call the latest appimage with --mode=http-proxy everything works very well.

Ok. But in your venv, i assume you had done *git pull* and then *pip install .* again ?

**Comment by pinnaculum** [comment]
I've tried with python 3.11.6 in both modes, with Lagrange, no problem whatsoever. Could you please show the output of *pip list* ?

**Comment by norayr** [comment]
yes, that's why i see something changed.
now i get python exceptions in lagrange.

before the problem was in my console.

in lagrange the configuration which works with appimage was not changed:
 it's just HTTP Proxy and it just localhost. without port.

i assumed that levior python binary and levior appimage would work the same way.

**Comment by norayr** [comment]
tried with localhost:1965 and it's the same.

**Comment by norayr** [comment]
output of pip list is

```
Package        Version Editable project location
-------------- ------- ----------------------------
aiogemini      0.2.0
aiohttp        3.9.0
aiohttp-socks  0.8.4
aiosignal      1.3.1
async-timeout  4.0.3
attrs          23.1.0
beautifulsoup4 4.12.2
cffi           1.16.0
CJKwrap        2.2
cryptography   41.0.5
diskcache      5.6.3
easydict       1.11
frozenlist     1.4.0
idna           3.4
levior         1.2.0   /home/user/pip/levior/levior
markdownify    0.11.6
md2gemini      1.9.1
mistune        2.0.5
multidict      4.7.6
pip            23.3.1
pycparser      2.21
python-socks   2.4.3
PyYAML         6.0.1
setuptools     68.2.2
six            1.16.0
soupsieve      2.5
wcwidth        0.2.10
wheel          0.41.2
yarl           1.9.2
```

**Comment by pinnaculum** [comment]
Thank you, the packages list seems OK. Could you please paste here the full exception that you see in Lagrange ? Is it an HTTPs URL that you're trying to access ?

**Comment by norayr** [comment]
```
$ ./levior --mode=http-proxy
levior: Built gemini service handler for: localhost
Processing request Request(url=URL('https://sdf.org/'))
Processing request Request(url=URL('https://sdf.org/?https://sdf.org'))
Processing request Request(url=URL('gemini://localhost/https:/sdf.org'))
```
i see proper requests in console, true. but the explosion of python happens and i see the output in lagrange.

by the way i don't understand, if i run levior with --mode=http-proxy why does it tell me to use gemini://localhost/https:/sdf.org ?

this is the whole log of the exception:

```
Traceback (most recent call last):
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1169, in _create_direct_connection
 hosts = await asyncio.shield(host_resolved)
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 880, in _resolve_host
 addrs = await self._resolver.resolve(host, port, family=self._family)
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/resolver.py", line 33, in resolve
 infos = await self._loop.getaddrinfo(
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/lib/python3.11/asyncio/base_events.py", line 867, in getaddrinfo
 return await self.run_in_executor(
	^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/lib/python3.11/concurrent/futures/thread.py", line 58, in run
 result = self.fn(*self.args, **self.kwargs)
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/lib/python3.11/socket.py", line 962, in getaddrinfo
 for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
socket.gaierror: [Errno -2] Name or service not known

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
 File "/home/user/pip/levior/levior/levior/handler.py", line 112, in handle_request_server_mode
 resp, rsc_ctype, rsc_clength, data = await crawler.fetch(
	^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/levior/levior/crawler.py", line 48, in fetch
 raise err
 File "/home/user/pip/levior/levior/levior/crawler.py", line 28, in fetch
 async with session.get(url, headers=headers,
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/client.py", line 1187, in __aenter__
 self._resp = await self._coro
	^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/client.py", line 574, in _request
 conn = await self._connector.connect(
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 540, in connect
 proto = await self._create_connection(req, traces, timeout)
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 907, in _create_connection
 _, proto = await self._create_direct_connection(req, traces, timeout)
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/home/user/pip/levior/lib/python3.11/site-packages/aiohttp/connector.py", line 1183, in _create_direct_connection
 raise ClientConnectorError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host https:80 ssl:default [Name or service not known]

```

**Comment by pinnaculum** [comment]
> ```
> $ ./levior --mode=http-proxy
> levior: Built gemini service handler for: localhost
> Processing request Request(url=URL('https://sdf.org/'))
> Processing request Request(url=URL('https://sdf.org/?https://sdf.org'))
> Processing request Request(url=URL('gemini://localhost/https:/sdf.org'))
> ```

Now i see what's wrong here. When you run levior in proxy mode, you shouldn't see a "Built gemini service handler for: localhost" message, you should see "levior: Serving in http proxy mode". You run by using *./levior*, is *./levior* the latest AppImage ?

I think i know what's happening. Although you're explicitely setting the mode with **--mode** on the command line, and even though you're not using *-c* to specify a config file, you have a *levior.yaml* in the current directory which sets *mode* to *server*, and levior (incorrectly) picks up the config file and ignores the mode passed on the command line. That's the only possible explanation.

Edit *levior.yaml* and set *mode* to *http-proxy* and this should fix the problem ! I'll fix this issue in the meantime, sorry for the inconvenience.

**Comment by norayr** [comment]
hello, so apparently i didn't know you asked for a full log, i just pasted the full log, and i didn't know we were online simultaneously.

so no, i was running python program levior

```
./levior --mode=http-proxy
levior: Built gemini service handler for: localhost
```

i use the appimage with the full name to know it is an appimage:

```
$ ./levior-latest-x86_64.AppImage --mode=http-proxy
levior: Serving in http proxy mode
```

yes, i did as you suggested:

```
$ mv levior.yaml /tmp/
./levior --mode=http-proxy
levior: Serving in http proxy mode
Processing request Request(url=URL('https://sdf.org/'))
```

and it works very well!

thank you!

**Comment by pinnaculum** [comment]
@norayr The issue has been fixed on master with v1.2.1, cmdline settings and config file settings (if a config file is passed) are merged, you can run it as a daemon now. In the next release *http-proxy* will be the default mode.

**Comment by norayr** [comment]
ty!
